### ISSUE TITLE
[Proposal] Initial / Original value API for AbstractControl

### ISSUE THREAD

**Issue by mkurcius** [angular/angular#19747]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
Right now form controls do not remember its original value.

## Expected behavior
FormControl, FormGroup. FormArray:
* has property `readonly original: any`
_holding original value_
* has property `readonly differ: boolean`
_holding `true` if value differ, `false` otherwise_
* has property `readonly valueDiffer: Observable<boolean>`
_stream indicating if current value differs from original_
* has method `revert()` / `rollback()`
_changes value to original value_
* has method `commit()`
_set existing value as new original value_

interface could looks like that:
```typescript
interface EnhancedAbstractControl {
  readonly original: any;
  readonly differ: boolean;
  readonly valueDiffer: Observable<boolean>;
  
  commit(): void;
  rollback(options?: { onlySelf?: boolean, emitEvent?: boolean });
}
```

## What is the motivation / use case for changing the behavior?
When creating form which updates current data it is also nice feature to allow user to revert changes for each individual field (instead of whole form) or set css class if value differ or show / hide other elements.

Sometimes also when user changes value from state A (original) -> B (end), some modal / warning need to be displayed now or when form is submitted or when value is changed. But for other values changes ex. A -> C or A -> D nothing happens. Validators are not candidates for this because value is valid.

With this new feature all this use cases will be easy to implement:
* rollback / revert
```html
<input [formControl]="control">
<button *ngIf="control.valueDiffer | async" (click)="control.rollback()">Undo</button>
```
* css
```html
<div [class.differ]="control.differ">
  <input [formControl]="control">
</div>
```
* before submit
```html
<form [formGroup]="form" (ngSubmit)="save()">
  <input formControlName="control">
</form>
```
```typescript
save() {
  const control = this.form.get('control');
  if (control.original === "A" && control.value === "B") {
    //ex. show modal
  }
}
```

**Comment by Toxicable** [comment]
>readonly differ: boolean;
  readonly valueDiffer: Observable<boolean>;

These fields can be derived trivially from just adding in `original` 
eg:
```
[class.differ]="control.original != control.value"
*ngIf="control.original != control.value"
```

>rollback(options?: { onlySelf?: boolean, emitEvent?: boolean });

this method is unneeded, just call `.setValue(control.original)`

I know these extra API's would make things slightly easier for the dev to use but from the core api aspect of it they're adding extra size to the bundle for no reason

___

Aside from the above I think this is a reasonable ask, there is currently no good ways to remember if the value has changed.
Current methods include:
* holding a mirrored object of the FormGroup/Control
* Extending FormControl

**Comment by mkurcius** [comment]
> I know these extra API's would make things slightly easier for the dev to use but from the core api aspect of it they're adding extra size to the bundle for no reason

Indeed these extra API's would make things easier for the developers and I thought angular was created for this reason :)

I agree with that a `rollback` method can be easily replaced with existing `setValue` or `reset` and dose not make much difference, but those 2 properties `differ` and `valueDiffer` are more valuable.

This solution:
```
[class.differ]="control.original != control.value"
*ngIf="control.original != control.value"
```
requires 2 checks of `control.original != control.value` and 2 bindings, 

when this:
```
[class.differ]="control.differ"
*ngIf="control.differ"
```
requires only 2 bindings (the value of differ can be calculated when setting new value in `updateValueAndValidity` and only return here) and looks more elegant

If the form will be larger then more unnecessary checks will be executed and app will be slower. So instead of slightly larger bundle (I am only guessing) we will end up with slightly slower app.


> Current methods include:
>* holding a mirrored object of the FormGroup/Control with just the original values to do a lookup
>* Extending FormControl

Yes, to achieve this behavior in my project, I extended each of FormControl, FormGroup and FormArray but end up with classes which are calling private methods of is base class, which does not looks good :/

**Comment by Toxicable** [comment]
>If the form will be larger then more unnecessary checks will be executed and app will be slower. So instead of slightly larger bundle (I am only guessing) we will end up with slightly slower app.

I disagree, logical comparison is a very fast operation, unless your form has several million of these then I don't think you will see any slow downs.

Also by not having the proposed fields you're actually saving perf on people who don't use them since it wont need to do the comparison when you're not using it.

___

One other thing I missed

>commit(): void;

I also this this is unneeded, the original value should never change, if you need to update some value you should use the FormControl's value then send that wherever you need it to go

**Comment by bracco23** [comment]
I have implemented a similar functionality, but instead of extending the FormControl class I have used a shadow deep copy of my model in the component where my form is. 

From the developer usability point of view, a good strategy could be to create a new directive that adds this functionality to the wanted FormControl, allowing the developer to decide which fields need the rollback feature and which not. The same directive, if applied to a FormGroup or a FormArray, would apply the feature to all the contained FormControls.

**Comment by mkurcius** [comment]
> I disagree, logical comparison is a very fast operation, unless your form has several million of these then I don't think you will see any slow downs.

Yes, I agree with you that logical comparison is fast. My point here was that it is unnecessary check in template, and it forcing developers to write uglier code. If form has 1-2 controls it may not looks so bad, but if there are 5+ controls it becomes to by ugly (it is of course my personal opinion).

As you said logical comparison is fast, so even if one won't use it, he / she won't notice difference in performance.

> One other thing I missed
>>commit(): void;

>I also think this is unneeded, the original value should never change, if you need to update some value you should use the FormControl's value then send that wherever you need it to go.
original is readonly so it cant be changed anyway

The idea behind `commit` is to handle situations like:
1. User submit form, request is send to server with OK response, but after that user can still see the form and can make another changes. Now original values should be the ones which returned with the response or some others (depends on business case) and developer should be able to mark them as new original values. Maybe `original` is misleading name here, maybe instead of `original` this field should be called `snapshot` or something similar.
The point here is to give user a experience that he / she change something, saved it and now he / she can change something again (maybe the same filed twice).
This experience can easily be achieved by adding class `differ` to  some element and changing background-color to orange.

2. The same situation as above but this time when submitting form server returns 409 (Conflict) or some other error. Error 409 may means optimistic locking exception and user will see modal with text "Inconsistent data" and buttons "Cancel" and "Refresh data". "Refresh data" button will re-fetch data and update form, from now on controls should have new `original` / `snapshot` values, so `commit` method should be called right after `setValues` to again user can experience that same as above.

Without `commit` it won't be possible, after form is updated (because of server response) controls will always be differ (have class .differ) and user won't be convinced every thing is ok.

**Comment by amoscatelli** [comment]
Why ? I think original property would be so much easier and friendly for programmers ...

**Comment by mkurcius** [comment]
For what purpose you need original value?

Initially I need for detecting changes in my object, by comparing prev value with current.
But I found it is too much verbose, especially with large objects or objects with dynamic keys.

Traversing FormGroup controls and filtering only dirty ones, turns out much simpler (and probably more performant)

**Comment by amoscatelli** [comment]
I neet to reset the control to the original value.
It would be simpler if it would be stored in the control itself as it's initialization value.

**Comment by jlem** [comment]
@mkurcius 

For what purpose you need original value? Isn't it obvious? You load up the edit form of a given piece of data (a customer, a recipe etc...) and you want to provide a user with the ability to reset the form to the original values after they start making changes. 

Currently, you have to manually track state to do this. Manual state tracking is obviously an error-prone way to write software and should be avoided when possible.

An API that abstracts away the need for the developer to manually track state would improve developer productivity - the fundamental purpose of a framework.

**Comment by mkurcius** [comment]
Hi, @jlem 
I was also struggling with similar problems (that's why I started this issue).
But it turns out it all depends on approach you choose.

In my case, approach that worked was to:
1. separate `container` and `presentational` component
2. `presentational` component receive `@Input` which is readonly model
3. based on model, create form (probably with it own "form" model)
4. now you can reset form anytime, using model from `@Input`
5. if user submits form:
    * emit event from `presentional` componet
    * handle it in `container` component and do your logic, like backend request etc.
    * when finished pass new value to `@Input` model
    * `presentional` component in `ngOnChnages` can now reset form based on new model :)

This approach also allows you to use `OnPush` change detection :)

If your problem dosen't meet with this approach, feel free to describe it in more details here and we can try to work out another solution :)

**Comment by hisham** [comment]
+1. Reset should reset to initial passed state. Otherwise don't let us pass in initial state when initializing a FormControl! :)

**Comment by SocVi100** [comment]
I think this feature is extremely useful. In fact I've used it extensively through a fantastic directive in AngularJS, but I haven't found a good replacement for it in Angular:

https://github.com/betsol/angular-input-modified

As stated in the info page of the project:

This Angular module adds additional properties and methods to the ngModel and ngForm controllers, as well as CSS classes to the underlying form elements to provide end-user with facilities to detect and indicate changes in form data.
This extra functionality allows you to provide better usability with forms. For example, you can add decorations to the form elements that are actually changed. That way, user will see what values has changed since last edit.
Also, you can reset an entire form or just a single field to it's initial state (cancel all user edits) with just a single call to the reset() method or lock new values (preserve new state) just by calling overloaded $setPristine() method.

**Comment by BlindDespair** [comment]
Is this going to be done? For a long time, I've been thinking that the way reset works is that it resets to the currently committed value. This is actually how native form works with `<button type="reset">`. I am convinced that the form should be able to remember initialValue, but it also should provide a way to commit for sure, it should not be readonly. The reason for that is when you save the changes you don't want to recreate your form completely, but instead, just say something like `.setValue()` `.commit()`. Another use case is when you initialize your form, but the current value is loaded asynchronously. The in `ngOnChanges` you could also do `.commit()` in order to let the form know that this is the current original value. 
I hope this will finally make it to the actual form API because IMO it's such a basic thing that almost every project using form could use.

**Comment by dylhunn** [comment]
Hi all! We spent some time today reviewing this issue. We think this boils down to two key missing features:

1. The ability to `reset` a `FormControl` to the initial provided value
2. The ability to read the default value on a control at any time

Both of these features will be provided in PR #44434.

Some things we are *not* doing:

- We don't want to add these to `AbstractControl` generally, only `FormControl`, since outer groups and arrays do not directly manage the reset behavior of inner controls. 
- We don't want to provide the ability to alter the default value at runtime -- this is too dynamic for our tastes.

**Comment by dylhunn** [comment]
This was fixed in #44434. Eventually, reset-to-initial-value will become the default behavior.
